pubtools/_quay/push_docker.py

```diff
@@ -415,26 +416,77 @@ def run(self):
             container_signature_handler = ContainerSignatureHandler(
                 self.hub, self.task_id, self.target_settings
             )
-            container_signature_handler.sign_container_images(docker_push_items)
+            manifest_claims, _ = container_signature_handler.sign_container_images(
+                docker_push_items
+            )
             # Push container images
             container_pusher = ContainerImagePusher(docker_push_items, self.target_settings)
             container_pusher.push_container_images()
 
             if operator_push_items:
                 # Build index images
                 operator_pusher = OperatorPusher(operator_push_items, self.target_settings)
+                existing_index_images = operator_pusher.get_existing_index_images(
+                    self.quay_api_client
+                )
                 iib_results = operator_pusher.build_index_images()
                 # Sign operator images
                 operator_signature_handler = OperatorSignatureHandler(
                     self.hub, self.task_id, self.target_settings
                 )
-                operator_signature_handler.sign_operator_images(iib_results)
+                operator_manifest_claims, _ = operator_signature_handler.sign_operator_images(
+                    iib_results
+                )
                 # Push index images to Quay
                 operator_pusher.push_index_images(iib_results)
         except Exception as e:
             LOG.error("An exception has occurred during the push, starting rollback")
             self.rollback(backup_tags, rollback_tags)
             raise e
+        else:
+            # Remove old signatures
+
+            new_signatures = [
+                (m["manifest_digest"], m["docker_reference"]) for m in manifest_claims
+            ]
+            outdated_signatures = []
+
+            for image_data, manifest in backup_tags.items():
+                ext_repo = image_data.repo.split("/")[1].replace(INTERNAL_DELIMITER, "/")
+                if "manifests" in manifest:
+                    for arch_manifest in manifest["manifests"]:
+                        outdated_signatures.append((arch_manifest, image_data.tag, ext_repo))
+                else:
+                    outdated_signatures.append((manifest["digest"], image_data.tag, ext_repo))
+
+            signatures_to_remove = []
+            for esig in container_signature_handler.get_signatures_from_pyxis(
+                [sig[0] for sig in outdated_signatures]
+            ):
+                if (
```

Comment:
Can you explain the purpose of this check? To see if signatures that should exist actually exist? It would make sense if we actually mapped unique signature IDs to the required fields

Reply:
existing signatures are fetched only by digests, so in return could be also signatures referring another repo (or tags). And second part of course checks if existing signatures isn't in "new" signatures which meant to be added in the current push